### ISSUE TITLE
No Easy Way To Create Instance of Class from Plain Object

### ISSUE THREAD

**Issue by rufio620** [bridgedotnet/Bridge#2389]
Is there any easy way to instantiate a bridge class from a JSON representation of that object. Say for example data coming back from a rest call which needs to be converted to the actual class type. Basically the opposite of Bridge.Script.ToPlainObject<T>. I know there is the ObjectLiteralAttribute but I want the objects treated as a given class based on the data returned from the server.

My first thought was to just use activator to create an instance of the object and them call bridge.merge on the object literal. However Bridge.merge does not seem to be exposed to the C# api so I am not sure how safe it is to use. This is also throws and exception (see deck below).

Ideally it would be nice to have a method Bridge.Script.FromPlainObject`<T>`(object objectLiteral) T;

### Steps To Reproduce

http://deck.net/e3f7883b0f29022b44038c67e7c920de

### Expected Result

ObjectLiteral Data should be merged into the class instance.

### Actual Result

```
Uncaught Error: division by zero
    at i.n.divide (bridge.min.js?15.7.0.1:8)
    at i.n.toString (bridge.min.js?15.7.0.1:8)
    at Object.merge (bridge.min.js?15.7.0.1:8)
    at Object.merge (bridge.min.js?15.7.0.1:8)
    at Object.merge (bridge.min.js?15.7.0.1:8)
    at $main (RunHandler.ashx?h=-1231077880:23)
    at HTMLDocument.i (bridge.min.js?15.7.0.1:8)
    at i (jquery-2.2.4.min.js:2)
    at Object.add [as done] (jquery-2.2.4.min.js:2)
    at n.fn.init.n.fn.ready (jquery-2.2.4.min.js:2)
```

**Comment by Zaid-Ajaj** [comment]
There is a way. 
When you serialize the object of type T on the server, send a string **x** along with the data representing the full type name such that eval(x) == Activator.CreateInstance(typeof(T)). 

```c#
// serialize type name
string x = GetEvaluatableTypeName<T>();
// create an instace
var instance = Script.Eval<T>(x);
```

Something like [this](http://deck.net/5507d983274300426f4097df4ff2fd15)

**Comment by vladsch** [comment]
We are going to merge the following PR with the master
https://github.com/bridgedotnet/Bridge/pull/2302

After that, proper serialization/deserialization will be possible in Bridge

Please note that class must be Reflectable

**Comment by rufio620** [comment]
I'm not sure this solves my problem. My example may have over simplified. I am retrieving data from and indexeddb collection. The data gets returned as an array of plain javascript object. I need a way to turn this into the class based on a generic type argument. I still don't see how to merge an existing data object into an instance of the corresponding class.

E.g. Based on the example above. I have the following js object

`var fromIndexedDb = { Name: "SomeName" } `

```c#
public class Person 
{
    public string Name { get;set;}
}
ConvertTo<T>(object objLit): T
{
   //how to convert objLit into Person Class
}
```

**Comment by Zaid-Ajaj** [comment]
@rufio620 Populating the `Person` instance with data from the object literal is as simple as reading the properties from the object literal one by one. 

Assume the `Person` class has the following properties
```c#
public class Person 
{
    public string Name { get;set;}
    public bool Married {get;set;}
    public int Age {get;set;}
}
```

and you have this object literal that has the data
```js
// Javascript
var person = { Name: "John", Married : true, Age: 20 };
```

Then you populate the data as follows
```c#
// csharp
// get an instance using this hack!
var personIntance = Script.Eval<Person>("new Demo.Person()");

// Populate the instance
personInstance.Name = person["Name"].As<string>();
personInstance.Married = person["Married"].As<bool>();
personInstance.Age = person["Age"].As<int>();
```

Example [here](http://deck.net/03dc4ad29dac22b55dab821e303ea5c9)

Note that this line 
```c#
var personIntance = Script.Eval<Person>("new Demo.Person()");
```
Can be replaced with the more C#-ish if you have type Person on the client :
```c#
// C# on the client
// [Reflectable] not required
var personIntance = Activator.CreateInstance<Person>();
```

**Comment by rufio620** [comment]
I think there is still some miscommunication on what I am asking. I understand what you are proposing, but what I am asking for is a way to do this with generics. With the solution proposed it assumes my function knows what concrete type it is operating on. If I have just a generic type T and need to merge data into a dynamically instantiated object of that type, there does not seem to be a clean way of doing this. 

E.g.
```js
// Javascript
var fromIndexedDb = { Name: "John", Married : true, Age: 20 };
```
```c#
var bridgeObj = Activator.CreateInstance<T>();
//how do I merge fromIndexedDb properties into bridgeObj. T is not known at design time.

```

**Comment by Zaid-Ajaj** [comment]
@rufio620 Then let me try again :) I understand you want the generic solution. 

Let's review how Bridge does things. If you have a class with PROPERTIES, then these properties become getter and setter METHODS. so if have you property `public string Name { get; set;}`, then two functions will be available `setName` and `getName` on the instance at run-time. For populating the instance with daata, we only need `set{PropertyName}` methods.

if they KEYS of the object literal  match with the names of properties then you have to ITERATE over every key, get the value from the object literal, invoke the `set{Key}` method to populate your instance. 

This method does that
```c#
public static T CreateInstance<T>(object data)
{
    var instance = Activator.CreateInstance<T>();
    foreach(var key in Object.Keys(data))
    {
        // read value from data by key
        var value = data[key];
       // set{PropertyName} is a function
       // thats why the .As<Action<object>>() part
       // we just invoke it
        instance["set" + key].As<Action<object>>()(value);
    }
    
    return instance;
}
```

[Here the sample](http://deck.net/f3cde395e79889f104363176ebc08ac4)

Hope this helps!

**Comment by michaelcheers** [comment]
You could make a template function which calls Bridge.merge or use JSON.Parse.

**Comment by geoffreymcgill** [comment]
If you have a json string of object, then `JSON.Parse<T>` might help, see

http://deck.net/0a0c3571a749d19d7400d1bce3318028

Please note, some changes to `JSON.Parse<T>` will be happening in the next release. The functionality is being worked on right now. See #2302.

**Comment by rufio620** [comment]
I think the solution posted by @Zaid-Ajaj will work. Thanks for all your help.
